src/H5Ocache.c

```diff
@@ -1268,6 +1268,11 @@ H5O__chunk_deserialize(H5O_t *oh, haddr_t addr, size_t chunk_size, const uint8_t
         UINT16DECODE(chunk_image, mesg_size);
         if (mesg_size != H5O_ALIGN_OH(oh, mesg_size))
             HGOTO_ERROR(H5E_OHDR, H5E_CANTLOAD, FAIL, "message not aligned");
+        if (mesg_size > (size_t)(p_end - chunk_image))
+            HGOTO_ERROR(H5E_OHDR, H5E_BADVALUE, FAIL, "message size exceeds buffer end");
+        if (id == H5O_MTIME_NEW_ID) /* validate mtime message size */
```

Comment:
I don't think there should be a specific case for mtime messages here. IF we really want to detect invalid messages we should create a callback for checking them, though I'm not sure we'll be able to detect all problems without fully decoding them, potentially adding unnecessary overhead.

There is a check for buffer overflow in H5O__mtime_new_decode(), why does this not catch the problem before an invalid access?

Reply:
It seems likely that the problem was with the off by one error in the check I mentioned in the other comment. I suspect things will work fine if you remove this one (if (id == H5O_MTIME_NEW_ID)), remove the one I mentioned (Try to detect invalidly formatted object header message that extends past end of chunk.), and implement Jordan's suggestions for the last one.